### PR TITLE
vendor and vendorLocation

### PR DESCRIPTION
**Pull request by OlegKandalov**
vendorLocation:
- mapper
- repo
- service
- controller
vendor:
- controller

### CODE REVIEW

File: src/main/java/com/exadel/discount/platform/converter/VendorLocationMapper.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import com.exadel.discount.platform.model.dto.VendorLocationDto;
+import org.modelmapper.ModelMapper;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Component
+public class VendorLocationMapper {
+    ModelMapper modelMapper;
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/converter/VendorLocationMapper.java:13]: private

File: src/main/java/com/exadel/discount/platform/web/VendorController.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    @GetMapping(value = "/{id}", produces = MediaType.APPLICATION_JSON_VALUE)
+    @PreAuthorize("hasRole('ADMINISTRATOR')")
+    public ResponseEntity<VendorDto> getVendorById(@PathVariable UUID id) {
+        VendorDto vendorDto = vendorService.getById(id);
+        return new ResponseEntity<>(vendorDto, HttpStatus.OK);
+    }
+
+    @PostMapping(value = "/add",
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/web/VendorController.java:34]: remove `value = "/add",`. Post method is used for creation so it's obvious that you're creating a new vendor

File: src/main/java/com/exadel/discount/platform/web/VendorController.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            consumes = MediaType.APPLICATION_JSON_VALUE,
+            produces = MediaType.APPLICATION_JSON_VALUE
+    )
+    @PreAuthorize("hasRole('ADMINISTRATOR')")
+    public ResponseEntity<VendorDto> createVendor(@RequestBody VendorDto vendorDto) {
+        VendorDto savedVendor = vendorService.save(vendorDto);
+        return new ResponseEntity<>(savedVendor, HttpStatus.CREATED);
+    }
+
+    @PutMapping(value = "/update",
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/web/VendorController.java:44]: should be`value = "/{id}"`. Put method is used for update so it's obvious that you're updating a vendor and you need to specify the id of the resource  which you're updating.

File: src/main/java/com/exadel/discount/platform/web/VendorLocationController.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    @PutMapping(value = "/update",
+            consumes = MediaType.APPLICATION_JSON_VALUE,
+            produces = MediaType.APPLICATION_JSON_VALUE)
+    @PreAuthorize("hasRole('ADMINISTRATOR')")
+    public ResponseEntity<VendorLocationDto> updateLocation(@PathVariable UUID id,
+                                                            @RequestBody VendorLocationDto locationDto) {
+        VendorLocationDto updatedLocation = locationService.update(id, locationDto);
+        return new ResponseEntity<>(updatedLocation, HttpStatus.OK);
+    }
+
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/web/VendorLocationController.java:52]: same as for vendorController

File: src/main/java/com/exadel/discount/platform/service/VendorLocationServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        return mapper.entityToDto(location);
+    }
+
+    @Override
+    public VendorLocationDto update(UUID id, VendorLocationDto vendorLocationDto) {
+        VendorLocation location = locationRepository
+                .findById(id)
+                .orElseThrow(() -> new NotFoundException("Vendor with id " + id +
+                        " was not found"));
+        location = mapper.dtoToEntity(vendorLocationDto);
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/VendorLocationServiceImpl.java:47]: map `vendorLocationDto` onto already existing location and make sure not to update the id

File: src/main/java/com/exadel/discount/platform/service/VendorLocationServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                .orElseThrow(() -> new NotFoundException("Location with id " + id +
+                        " was not found"));
+        return mapper.entityToDto(location);
+    }
+
+    @Override
+    public VendorLocationDto update(UUID id, VendorLocationDto vendorLocationDto) {
+        VendorLocation location = locationRepository
+                .findById(id)
+                .orElseThrow(() -> new NotFoundException("Vendor with id " + id +
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/VendorLocationServiceImpl.java:45]: Vendor was not found?

File: src/main/java/com/exadel/discount/platform/service/VendorLocationServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        return mapper.entityToDto(location);
+    }
+
+    @Override
+    public void toArchive(UUID id) {
+        boolean isExists = locationRepository.existsById(id);
+        if (isExists) {
+            locationRepository.deleteById(id);
+        } else {
+            throw new NotFoundException("Location with id " + id +
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/VendorLocationServiceImpl.java:58]: we have 2 different Location entities in the project so be more specific  here

**Review by ohalushchak-exadel — CHANGES_REQUESTED** [review]

File: src/main/java/com/exadel/discount/platform/service/VendorLocationServiceImpl.java [hunk 8]
```diff
@@ -33,20 +35,20 @@ public VendorLocationDto save(VendorLocationDto vendorLocationDto) {
...
                 .findById(id)
-                .orElseThrow(() -> new NotFoundException("Location with id " + id +
+                .orElseThrow(() -> new NotFoundException("Vendor location with id " + id +
                         " was not found"));
         return mapper.entityToDto(location);
     }
 
     @Override
-    public VendorLocationDto update(UUID id, VendorLocationDto vendorLocationDto) {
+    public VendorLocationResponseDto update(UUID id, VendorLocationDto vendorLocationDto) {
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/VendorLocationServiceImpl.java:44]: use this DTO in Get/GetAll as well

File: src/main/java/com/exadel/discount/platform/converter/VendorLocationMapper.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+    public VendorLocationDto entityToDto(VendorLocation vendorLocation) {
+        return modelMapper.map(vendorLocation, VendorLocationDto.class);
+    }
+
+    public VendorLocation dtoToEntity(VendorLocationDto locationDto) {
+        return modelMapper.map(locationDto, VendorLocation.class);
+    }
+
+    public <S, T> List<T> mapList(List<S> source, Class<T> targetClass) {
```
[the remark this answers is not in the document]
Comment (OlegKandalov) [reply src/main/java/com/exadel/discount/platform/converter/VendorLocationMapper.java:36]: i took this method from the site: https://www.baeldung.com/java-modelmapper-lists 
))

**Review by ohalushchak-exadel — APPROVED** [review]

File: src/main/java/com/exadel/discount/platform/converter/VendorLocationMapper.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (nkharitoniuk) [reply src/main/java/com/exadel/discount/platform/converter/VendorLocationMapper.java:36]: It's just a suggestion to use a library for example https://stackabuse.com/guide-to-mapstruct-in-java-advanced-mapping-library

**Review by nkharitoniuk — APPROVED** [review]

**Pull request merged**